rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,829 @@
+resolve the nested name inside the compiler.
+
+### Explicit command line checking/substitution vs attributes: 
+
+Instead of
+  adding a new special `proof_for_contact` attributes we could have instead done:
+
+  1. **Check contracts on the command line** like CBMC does. This makes contract
+     checking a separate `kani` invocation with something like a
+     `--check-contract` flag that directs the system to instrument the function.
+     This is a very flexible design, but also easily used incorrectly.
+     Specifically nothing in the source indicates which harnesses are supposed
+     to be used for which contract, users must remember to invoke the check and
+     are also responsible for ensuring they really do verify *all* contacts they
+     will later be replacing and lastly.
+  2. **Check contracts with a `#[kani::proof]` harness.** This would have used
+     e.g. a `#[kani::for_contract]` attributes on a `#[kani::proof]`. Since
+     `#[kani::for_contract]` is *only* valid on a proof, we decided to just
+     imply it and save the user some headache. Contract checking harnesses are
+     not meant to be reused for other purposes anyway and if the user *really*
+     wants to the can just factor out the actual contents of the harness to
+     reuse it.
+
+### Polymorphism during contract checking
+
+A current limitation with how contracts are enforced means that if the target of
+a `proof_for_contract` is polymorphic, only one monomorphization is permitted to
+occur in the harness. This does not limit the target to a single occurrence,
+*but* to a single instantiation of its generic parameters.
+
+This is because we rely on CBMC for enforcing the `modifies` contract. At the
+GOTO level all monomorphized instances are distinct functions *and* CBMC only
+allows checking one function contract at a time, hence this restriction.
+
+### User supplied harnesses
+
+We make the user supply the harnesses for checking contracts. This is our major
+source of unsoundness, if corner cases are not adequately covered. Having Kani
+generate the harnesses automatically is a non-trivial task (because heaps are
+hard) and will be the subject of [future improvements](#future-possibilities). 
+
+In limited cases we could generate harnesses, for instance if only bounded types
+(integers, booleans, enums, tuples, structs, references and their combinations)
+were used. We could restrict the use of contracts to cases where only such types
+are involved in the function inputs and outputs, however this would drastically
+limit the applicability, as even simple heap data structures such as `Vec`,
+`String` and even `&[T]` and `&str` (slices) would be out of scope. These data
+structures however are ubiquitous and users can avoid the unsoundness with
+relative confidence by overprovisioning (generating inputs that are several
+times larger than what they expect the function will touch).
+
+
+## Open questions
```

Comment:
I actually have one suggestion that I believe should go in the open questions section for now:

Panic conditions. I do think that being able to specify, check and stub under which conditions the code should panic is very powerful, and it should be part of the contract of a rust function.

Reply:
One could argue that safe functions should always panic if the input is a valid value for the given type. And maybe Kani could assume that, and actually enforce this case.

If a valid value for the given type may cause UB, the function should be declared unsafe. In that case, maybe we could provide an attribute to specify panic conditions to differ from conditions that may lead to UB.

Reply:
Those are two interesting points but I think they're quite different.

To the first: usually you'd use a precondition to rule out a panic. But I've added this to open questions non the less because there could be situations in which a user may want to specify that panicking is expected behavior.

To the second: I agree this should be added.